### PR TITLE
Improve DynamoDB condition expression support

### PR DESCRIPTION
**Pull request by garrettheel**
Picking up @mjstevens777's work in https://github.com/spulec/moto/pull/2136.

This adds proper support for parsing and evaluating condition expressions.

### CODE REVIEW

File: tests/test_dynamodb2/test_dynamodb.py [hunk 1]
```diff
@@ -1641,6 +1718,89 @@ def test_condition_expressions():
...
+    client.put_item(
+        TableName='test',
+        Item={
+            'forum_name': {'S': 'foo'},
+            'ttl': {'N': 'bar'},
+        }
+    )
+
+
+    def update_if_attr_doesnt_exist():
```
Comment (garrettheel) [inline tests/test_dynamodb2/test_dynamodb.py]: @killthrush I think this captures your comment from #2136 but let me know if not!

Comment (killthrush) [reply tests/test_dynamodb2/test_dynamodb.py]: Thanks for picking this up @garrettheel!  This fixes one issue, however in the interim I've collected a few more test cases related to conditional checks that still fail with this commit.  I'll see if I can capture them too by cloning your branch.

Comment (killthrush) [reply tests/test_dynamodb2/test_dynamodb.py]: Hmm, it looks like some of these failures are unrelated to moto.  I'll keep testing, but feel free to move ahead with this PR - I can always file an issue after the fact.  Thank you!

Comment (killthrush) [reply tests/test_dynamodb2/test_dynamodb.py]: OK, I *did* find some test failures related to `list_append` not working properly, but I suspect that is a different issue judging from the commit history.

**Comment by garrettheel** [comment]
@terrycain I'd really appreciate if you could take a look and let me know what this needs to be merged. thanks!

**Comment by garrettheel** [comment]
@spulec would you mind taking a look?

**Comment by spulec** [comment]
Based on coveralls, it looks like this added a lot of new code that isn't covered by tests. Can you check into that?

**Comment by mjstevens777** [comment]
Sorry, it looks like I left some dead code in `moto/dynamodb2/responses.py`. Lines 35-93.

**Comment by garrettheel** [comment]
Removed the existing code that was made irrelevant by this change!

**Comment by spulec** [comment]
Awesome; thanks for all of your work!

**Comment by alexholtz** [comment]
I could be mistaken, but from what I can tell the way this was implemented it seems like the API-style conditions such as `attribute_not_exists('attr1')` is supported, but the SDK style conditions such as: `Attr('attr1').not_exists()` are not.

**Comment by alexholtz** [comment]
I could be mistaken, but it seems like the way this was implemented it supports the API-style ConditionExpressions: `ConditionExpression=attribute_not_exists('attr_name')` but not the SDK-style conditions: `ConditionExpression=Attr('attr_name').not_exists()`.

**Pull request merged**